### ISSUE TITLE
RFC: Helm chart support

### ISSUE THREAD

**Issue by alpeb** [linkerd/linkerd2#3041]
This is a list of possible routes to tackle supporting installing Linkerd through a Helm chart. Some of these aspects have been touted in other Github issues, but I wanted to get the discussion restarted, and in light of the state of Linkerd after 2.4.

Note this only applies to the installation of the control plane. The user pods are now injected automatically by default by the proxy injector, and we assume that's the common production scenario.

Desired qualities
=============
- The end result of installing through Helm should be identical to installing through `linkerd install`
- We don't want people to have to install extra stuff (e.g. a Helm plugin)

Possible solutions
==============
Note all these require generating at least the uninjected control plane chart for each release.

Fully injected chart, no hooks
------------------------------------------
For each release, generate the control plane chart, already injected (manually at first, then through some script). That's the approach taken in #2424 .
To be able to get the exact same result as with `linkerd install`, we need to translate all of the options validation go code into template helpers, as well of some of the injection code. Note that Helm already has template functions for generating the certs and private keys.

Non-injected chart plus injection hook
---------------------------------------------------
For each release, generate the control plane chart, uninjected. Have a post-install Helm hook that would run `linkerd inject --manual` on the deployments from the control plane (--manual so that the deployments themselves get injected and not the pods, like with `linkerd install`).
The hook needs to communicate with the cluster to retrieve the deployments from the control plane, and `linkerd inject` also needs such connection to fetch the linkerd-config. So we'd need to manufacture a kubeconfig file based off of the hook pod's service account token. And we'd have to setup the proper RBAC permissions to that SA in the hook's yaml itself.

Non-injected chart plus simple hook for auto-injection
-----------------------------------------------------------------------
For each release, generate the control plane chart, uninjected. Have a post-install Helm hook insert the right annotation in the control plane deployments for having the proxy-injection automatically do the injection. This would also require setting up the kubeconfig and permissions as in the previous solution.
There are many downsides to this solution:
- The injection happens at the pod-level, so the result would be different than `linkerd install`'s
- Currently the Identity pod doesn't work if uninjected, so that'd need to addressed
- The control plane currently explicitly opts-out of auto-injection, at least because of bootstrapping issues related to HA. Will have to address that.

Discarded solutions
================
- Helm plugin: it would be really easy to embed the CLI into a plugin, but that requires the extra step of installing the plugin
- Running `linkerd install` in a pre-install hook for it to take care of everything, and piping the resulting resources into k8s. The problem is Helm won't be aware of those resources and thus it can't manage them.

**Comment by grampelberg** [comment]
> Fully injected chart, no hooks

This feels like the best option right now. So far, the primary issue to going this route is how to have a single code path. Here's my thinking on a way that is achievable:

- Create a chart that just contains helper methods:
    - proxy container
    - debug container
    - metadata
    - volume 
- Create a control plane chart that:
    - Depends on the helper methods
    - Uses the partials in the correct locations
    - Outputs a fully injected control plane, `linkerd install` would become a simple wrapper of this.
- Create an "inject" chart that:
    - Depends on the helper methods
    - Outputs json patch format
- Update `linkerd inject` to only use patch

This should actually clean up a decent amount of code and make sure there's only a single source of truth for injection.

**Comment by adleong** [comment]
`linkerd install` also does some checks/validation.  Is that in scope when using helm for install?  How would that work?

You mention that Helm has template functions for generating the certs and private keys.  Can you go into more detail here?  We generate our certs and keys in a pretty specific way, can Helm support that?

How does Helm interact with RBAC and user permissions?  Will all resources be created by an account with cluster admin permissions?  e.g. Helm can create cluster level resources?

Is a Helm plugin automatically off the table?  Adding an extra step of installing the plugin is not ideal, but if that makes the rest of the process easier, maybe we should evaluate that tradeoff.

**Comment by grampelberg** [comment]
> linkerd install also does some checks/validation. Is that in scope when using helm for install? How would that work?

We could *technically* do some pre-install hooks, but it wouldn't be a great experience. A chart will end up being `advanced` mode.

> You mention that Helm has template functions for generating the certs and private keys. Can you go into more detail here? We generate our certs and keys in a pretty specific way, can Helm support that?

It should, http://masterminds.github.io/sprig/crypto.html.

> How does Helm interact with RBAC and user permissions? Will all resources be created by an account with cluster admin permissions? e.g. Helm can create cluster level resources?

Today, most users run `tiller` that uses a service account with `cluster-admin`. This'll just end up being part of what we can't make better with helm. Tomorrow, helm v3 uses the local kubeconfig and it is a little more friendly.

> Is a Helm plugin automatically off the table? Adding an extra step of installing the plugin is not ideal, but if that makes the rest of the process easier, maybe we should evaluate that tradeoff.

Definitely not off the table from my perspective, I'm reasonably sure a plugin wouldn't help us much though as plugins are totally separate from the normal `install` flow.

**Comment by ihcsim** [comment]
I wonder if we can introduce a new install subcommand to generate [named templates](https://github.com/helm/helm/blob/master/docs/chart_template_guide/named_templates.md) for all the dynamic parts of the charts. And then Helm users will just use this subcommand to generate their own partial template files. E.g.
```
$ linkerd install partials
{{- define "linkerd.proxy.spec" -}}
- name: linkerd-proxy
  image: gcr.io/linkerd-io/proxy: edge-19.7.1
....
{{- end -}}
{{- define "linkerd.proxy-init.spec" -}}
- name: linkerd-init
  image: gcr.io/linkerd-io/proxy-init:v1.0.0
...
{{- end -}}
{{- define "linkerd.debug.spec" -}}
...
{{- end -}}
$ linkerd install partials > ./chart/templates/_partials.yaml`
$ helm install ./chart
```

We will have to update all existing charts to [include](https://github.com/helm/helm/blob/master/docs/chart_template_guide/named_templates.md#the-include-function) these partials. 

Code and CLI options sharing between `install` and `install partials` should be possible, even though it will probably require some refactoring.

Outside of Helm, `install` will treat the output of `install partials` as [`Template`](https://godoc.org/k8s.io/helm/pkg/proto/hapi/chart#Template)s, and add them to the slice of static charts prior to rendering.
```go
	// Create chart and render templates
	chrt, err := chartutil.LoadFiles(files)
	if err != nil {
		return err
	}

        // should come from some functions shared with `install partials`
	partials := &chart.Template{
		Data: []byte(`{{- define "linkerd.proxy.spec" -}}
- name: linkerd-proxy
  image: gcr.io/linkerd-io/proxy: edge-19.7.1
....
{{- end -}}
{{- define "linkerd.proxy-init.spec" -}}
- name: linkerd-init
  image: gcr.io/linkerd-io/proxy-init:v1.0.0
...
{{- end -}}
{{- define "linkerd.debug.spec" -}}
...
{{- end -}}
`),
	}
	chrt.Templates = append(chrt.Templates, partials)

	renderOpts := renderutil.Options{
		ReleaseOptions: chartutil.ReleaseOptions{
			Name:      "linkerd",
			IsInstall: true,
			IsUpgrade: false,
			Time:      timeconv.Now(),
			Namespace: controlPlaneNamespace,
		},
		KubeVersion: "",
	}
```
Not sure if this will help; might be worth a spike? At least we won't have to deal with hooks and RBAC.

**Comment by grampelberg** [comment]
@ihcsim it sounds like we're pretty much on the same page. Why would we need `install partials` at all?

**Comment by alpeb** [comment]
Following up on the idea to generate the the chart dynamically and then feeding it into `helm install .`, and If a Helm plugin is not off the table, we can do this through a plugin configured like so:

```bash
$ tree helm/linkerd/
helm/linkerd/
├── chart
│   ├── Chart.yaml
│   └── templates
├── install
├── linkerd-install.sh
└── plugin.yaml
```

plugin.yaml:
```yml
name: "linkerd"
version: "edge-19.7.2"
usage: "helm linkerd install|upgrade [flags]"
description: "This is a Helm plugin wrapper for the linkerd CLI"
ignoreFlags: false
useTunnel: false
command: "$HELM_PLUGIN_DIR/linkerd-install.sh"
hooks:
  # this is the same as https://run.linkerd.io/install 
  install : "$HELM_PLUGIN_DIR/install
  upgrade: "$HELM_PLUGIN_DIR/install
```

linkerd-install.sh:
```bash
#!/bin/sh

cd /home/alpeb/go/src/github.com/linkerd/linkerd2/helm/linkerd
# TODO: forward flags
$HOME/.linkerd2/bin/linkerd install > chart/templates/install.yaml
cd chart
helm install .
```

The workflow would then be:
```bash
# install the plugin
helm plugin install https://run.linkerd.io/helm-plugin

# run it
helm linkerd [flags]
```

The above creates a new Helm release, that can be deleted with `helm delete`.
This just caters the install scenario; some additional bash foo would be required to support upgrades (thus allowing `helm rollback`, etc)

**Comment by grampelberg** [comment]
@alpeb the plugin proposal just makes `linkerd install` a subcommand of `helm`?

**Comment by alpeb** [comment]
> the plugin proposal just makes `linkerd install` a subcommand of `helm`?

@grampelberg Yes, it would be a subcommand to be called through `helm linkerd`, but it wouldn't be just a simple wrapper, as under the hood it would first generate the chart dynamically using `linkerd install` and then it would call `helm install` pointing to the generated chart.

Also, `helm linkerd` would install the linkerd CLI as a first step.

**Comment by grampelberg** [comment]
I was trying to figure out how users would customize the install.

Can a plugin be a dependency from another chart? A pretty common use case for helm is creating an “umbrella” chart that defines the entire environment through requirements.yaml (see the [workflow chart](https://github.com/deis/workflow/tree/master/charts/workflow)).

**Comment by alpeb** [comment]
My plugin proposal was very simplistic in that you would customize the install by passing linkerd flags into `helm linkerd` (you could list those flags with `helm linkerd -h`).

If instead of that we'd like to support customization through something like `values.yml`, then we could add a new `-f` flag that would convert such file into the appropriate options.

Alternatively, as mentioned elsewhere in this ticket, we could move the go logic into our templates to have Helm itself do the placeholder replacement, but the more I think about that, the more it scares me. Also as pointed by @adleong, by not going through the CLI we loose validation.

@grampelberg I didn't understand your point about dependencies, could you clarify?

**Comment by grampelberg** [comment]
> @grampelberg I didn't understand your point about dependencies, could you clarify?

Take a look at the chart that I linked, there’s no templates in the main chart - everything is a dependency.

> Alternatively, as mentioned elsewhere in this ticket, we could move the go logic into our templates to have Helm itself do the placeholder replacement, but the more I think about that, the more it scares me.

I believe my proposal does this without any scariness. Is there something lacking there?

> Also as pointed by @adleong, by not going through the CLI we loose validation.

We should always promote `linkerd install` as the primary method of installation. I don’t think we should worry about creating an identical experience with the helm install. It should be positioned as an advanced method and if you shoot your foot off, you came into it with awareness (we’re no different than the rest of the k8s ecosystem with helm there).

**Comment by ihcsim** [comment]
> Can a plugin be a dependency from another chart?
> @grampelberg I didn't understand your point about dependencies, could you clarify?

IOW, will the Linkerd Helm plugin still work, if the Linkerd chart is a subchart of the umbrella chart? Or will the user need to install the Helm plugin in the parent chart?

> I don’t think we should worry about creating an identical experience with the helm install. 

+1 on this one.

@grampelberg IIUC, the difference between what you and I proposed is that I am hoping to not have to create static charts with helper methods for proxy container, debug container, metadata and volumes. Instead, we dynamically generate these charts in Go. Since we already have Go code that knows how to generate these YAML, it should be easy to output them as named templates. And the reason for `linkerd install partials` is to provide users a way to output these dynamically generated templates, and pipe them into files. Once these named templates are saved in files, users can resume whatever Helm workflows they already have in-place.

**Comment by grampelberg** [comment]
> And the reason for linkerd install partials is to provide users a way to output these dynamically generated templates, and pipe redirect them into files.

I still don’t understand this one. `linkerd install partials` creates the helper chart that I’m suggesting but requires all configuration to happen from the CLI instead of a `values.yaml` file? I guess I’m unclear on what this gets us.

My understanding of the situation is that the placement of the partials is the tough bit, not the actual generation. Does this proposal cover the `install` flow, but doesn’t address the `inject —manual` and auto-inject code paths?

**Comment by alpeb** [comment]
So I'm exploring moving some of the logic into the templates. One hurdle I found is that the `genSelfSignedCert` template function from Helm gives RSA keys instead of EC. Dunno if we wanna make that switch. Also I guess that implies changing things in the proxy... 

For the proxy-injector's webhook config `caBundle`, I've seen this addressed in other projects by setting it to an empty string, and generating the bundle and updating it after the injector starts. So similar to what we used to do when we created the config at runtime, but now we'd only have to update the `caBundle` properties (also applies to the sp-validator).

We could do the same for the indentity pod (unless the user provides its own).

**Comment by alpeb** [comment]
... note that the change we made to create the webhook configs at install time was to support HA, so that if we had multiple proxy-injector pods we wouldn't have a race between them trying to set up the webhook config. So if we now are going to update the config at runtime we have to figure some locking mechanism.

**Comment by grampelberg** [comment]
> One hurdle I found is that the genSelfSignedCert template function from Helm gives RSA keys instead of EC.

We should probably get a decision from @olix0r on this. Outside of that, I’m content either:

- Requiring the keys to be generated outside of helm and passed in as values.
- Special helm pre-install hook that generates the certs. As we’ll just be taking care of that in `linkerd install`, there’s nothing special required.

> For the proxy-injector's webhook config caBundle, I've seen this addressed in other projects by setting it to an empty string, and generating the bundle and updating it after the injector starts.

No need to do this, I’ve got it working in [smi-metrics](https://github.com/deislabs/smi-metrics/blob/master/chart/templates/apiservice.yaml).

**Comment by alpeb** [comment]
@grampelberg Like in your smi-metrics example, for the proxy-injector and the sp-validator we should get away with `genSignedCert`. It would require some minor changes on our end to adapt to RSA certs.

As for the root cert/key for the identity pod, I like the pre-install hook approach (which would be a no-op in the case the user provides that cert/key).

**Comment by alpeb** [comment]
btw, to answer the point about the helm plugin as a chart dependency, I believe that's not possible, given plugins are a client-side thing that are setup through `helm plugin install` only.

I'll continue exploring the charts templates/partials enhancements and report back.

**Comment by grampelberg** [comment]
> It would require some minor changes on our end to adapt to RSA certs.

Ahh, I didn’t realize, that’s good to know.

**Comment by alpeb** [comment]
@grampelberg I'm going now through the injection part. You suggested having an "inject" chart that would output a json patch. How would we apply that patch? AFAIK that's not implemented directly yet in Helm (https://github.com/helm/helm/issues/3181). 
Should we apply it with `kubectl patch` in a post-install hook?

**Comment by alpeb** [comment]
After some offline conversation, here's a preliminary list of things that we can focus on. Let me know what you think.

JSON patch-related changes:
- Create a new helm chart just for the proxy container (consider if a simple go template can do the trick)
- For `linkerd install` and the webhook proxy-injector: Refactor `injectPodSpec()` in `pkg/inject/inject.go` that currently generates a JSON patch, but have it use the proxy container chart instead of the hard-coded go-client structs.

Once that approach is validated:
- Create chart for init-container
- Refactor `injectProxyInit` to use that chart.

The following can be started in parallel to the previous tasks:
- Refactor the `proxy-injector` and `sp-validator` charts so that they generate the certs/keys with helm's `genSelfSignedCert()`.
- Create `values.yml` with all the default values, by hand (later, we can have this be automated based off of protobuf for the config part). The trust root for identity is expected to be provided by the user in this file.
- Refactor the current charts so that they rely on the charts mentioned above, so that `helm install` can produce a fully injected control plane.

**Comment by grampelberg** [comment]
I’ve got a couple questions, everything else looks great. The external trust root should *definitely* be called out as part of the design process and it might surprise folks.

> Create chart for init-container

Do you mean template? I think the patch chart will actually work out best as a single template containing everything as the patch should be a single document.

> Refactor injectProxyInit to use that chart.

As I understand your explanation (please correct my assumptions if I’m wrong), `injectPodSpec()` and `injectProxyInit()` would both be folded into a single function.

**Comment by ihcsim** [comment]
@grampelberg To confirm my understanding on https://github.com/linkerd/linkerd2/issues/3041#issuecomment-508860859, does the following way of organizing the charts sound reasonable?

* Introduced a new `charts` parent folder, containing 3 charts folders:
   * the `partials` chart has partial templates which contain all the dynamic portions (proxy, proxy init, volumes, config map etc.) of the Linkerd chart.
   * the `control-plane` chart has templates for all the uninjected control plane components and other non-workload resources. It depends on the `partials` chart, and it's used by `linkerd install`.
   * the `data-plane` ("patch") chart has a single JSON patch template. It depends on the `partials` chart, and it's used by `linkerd inject --manual` and auto-inject. (`linkerd inject` will retain its currently function of adding annotations only.)

```sh
➜  linkerd2 git:(master) ✗ tree charts
charts
├── control-plane
│   ├── charts
│   │   └── partials -> ../../partials
│   ├── Chart.yaml
│   ├── templates
│   │   ├── config.yaml
│   │   ├── controller-rbac.yaml
│   │   ├── controller.yaml
│   │   ├── grafana-rbac.yaml
│   │   ├── grafana.yaml
│   │   ├── identity-rbac.yaml
│   │   ├── identity.yaml
│   │   ├── namespace.yaml
│   │   ├── NOTES.txt
│   │   ├── prometheus-rbac.yaml
│   │   ├── prometheus.yaml
│   │   ├── proxy_injector-rbac.yaml
│   │   ├── proxy_injector.yaml
│   │   ├── psp.yaml
│   │   ├── _resources.yaml # move this to partials?
│   │   ├── serviceprofile-crd.yaml
│   │   ├── sp_validator-rbac.yaml
│   │   ├── sp_validator.yaml
│   │   ├── tap-rbac.yaml
│   │   ├── tap.yaml
│   │   ├── trafficsplit-crd.yaml
│   │   ├── web-rbac.yaml
│   │   └── web.yaml
│   └── values.yaml
├── data-plane
│   ├── charts
│   │   └── partials -> ../../partials
│   ├── Chart.yaml
│   ├── templates
│   │   ├── NOTES.txt
│   │   └── patch.json
│   └── values.yaml
└── partials
    ├── charts
    ├── Chart.yaml
    ├── templates
    │   ├── _debug.yaml
    │   ├── _metadata.yaml
    │   ├── _proxy.yaml
    │   └── _volumes.yaml
    └── values.yaml
```

**Comment by grampelberg** [comment]
@ihcsim yeah!

**Comment by alpeb** [comment]
@ihcsim's layout looks good to me, barring proxy-init missing from partials, which looks like just an unintended omission. That should answer @grampelberg's questions above.
Regarding merging `injectPodSpec()` and `injectProxyInit()`, it does sound so, since the logic to whether produce an init container or not will belong into `patch.json`.

Is there anything else you'd like to discuss, or do you guys feel comfortable at this stage for us to kick-start development through a new umbrella ticket and corresponding sub-tickets? Or do you feel we wanna have a separate design doc?

**Comment by grampelberg** [comment]
I’d like to have a record of what we’re planning on doing. That could either be:

- This issue updated with the plan.
- A new issue with the plan, linked from here.
- A design doc that gets checked in somewhere and linked.

In all three instances, I’d love to have a short explanation of what we’re planning on doing, why we’re doing it and pointers to the locations that contain the reasons we’re not doing it another way.

In a perfect world, that’d include some user flows for how they’re expected to use it. Would you mind walking through that at a high level? I’m sure we’re not thinking about something and it’d help structure the inputs.

**Comment by alpeb** [comment]
The design document for Helm support was tracked in #3099 
The ensuing issues are being tracked under the [Helm support project](https://github.com/orgs/linkerd/projects/13).
